src/common/validators.ts

```diff
@@ -0,0 +1,243 @@
+
+  if (isEmpty(decoratorOptions)) {
+    throw new Error(logger.createMessage(`Invalid decorator options. It cannot be empty.`));
+  }
+
+  if (typeof decoratorOptions !== 'object' || Array.isArray(decoratorOptions)) {
+    throw new Error(logger.createMessage(`Invalid decorator options. It must be an object.`));
+  }
+
+  if (typeof options !== 'object' || Array.isArray(options)) {
+    throw new Error(logger.createMessage(`Invalid validator options. It must be an object.`));
+  }
+
+  const { allowUnknown = false, known = [], required = [] } = options;
+
+  if (known.includes('domain') && decoratorOptions.domain) {
+    validateDomain(decoratorOptions.domain);
+  }
+
+  if (known.includes('name') && (decoratorOptions.name || configuration.isStrict())) {
+    validateNameOrIdentifier(decoratorOptions.name);
+  }
+
+  for (const name of required) {
+    if (name in decoratorOptions && !isEmpty(decoratorOptions[name])) {
+      continue;
+    }
+    throw new Error(logger.createMessage(`Invalid decorator options. The option '${name}' is required.`));
+  }
+
+  if (allowUnknown) {
+    return;
+  }
+
+  const names = Object.getOwnPropertyNames(decoratorOptions);
+  const unknown = names.filter(name => !known.includes(name));
+
+  if (unknown.length > 0) {
+    throw new Error(logger.createMessage(`Invalid decorator options. The unknown options are not allowed: ${unknown.join(', ')}.`));
+  }
+}
+
+interface ValidateDomainOptions {
+  /**
+   * Allows the domain to denote hierarchy by separating chunks with a /.
+   *
+   * @default true
+   */
+  allowHierarchy: boolean;
+}
+
+/**
+ * Validates domain.
+ *
+ * @param domain the domain to be validated.
+ * @param options the options to customize the validation.
+ *
+ * @throws Error if the {@link domain} is blank.
+ * @throws Error if the {@link domain} is empty.
+ * @throws Error if the {@link domain} is a boolean.
+ * @throws Error if the {@link domain} is an array.
+ * @throws Error if the {@link domain} is an object.
+ * @throws Error if the {@link domain} has chunks and {@link options.allowHierarchy} is {@link false}.
+ * @throws Error if any of the {@link domain} chunks is blank.
+ * @throws Error if any of the {@link domain} chunks is empty.
+ * @throws Error if any of the {@link domain} chunks has spaces.
+ */
+function validateDomain(domain: any, options?: ValidateDomainOptions): void {
+  // logger.debug('validating domain:', domain);
+
+  const { allowHierarchy = true } = options ?? {};
+
+  if (isEmpty(domain)) {
+    throw new Error(logger.createMessage(`Invalid domain. It cannot be empty.`));
+  }
+
+  if (typeof domain === 'boolean') {
+    throw new Error(logger.createMessage(`Invalid domain. It cannot be a boolean.`));
+  }
+
+  if (typeof domain === 'object') {
+    if (Array.isArray(domain)) {
+      throw new Error(logger.createMessage(`Invalid domain. It cannot be an array.`));
+    }
+    throw new Error(logger.createMessage(`Invalid domain. It cannot be an object.`));
```

Comment:
Wouldn't it be easier to just check for string a throw if not?

Reply:
Where's the fun